dnscache/storage.go

```diff
@@ -0,0 +1,94 @@
+package dnscache
+
+import (
+	"net"
+	"reflect"
+	"time"
+
+	"fmt"
+
+	"github.com/pmylund/go-cache"
+)
+
+// DnsCacheItem represents single record in cache
+type DnsCacheItem struct {
+	addrs []string
+}
+
+func (item *DnsCacheItem) IsEqualsTo(addrs []string) bool {
+	return reflect.DeepEqual(item.addrs, addrs)
+}
+
+// DnsCacheStorage is an in-memory cache of auto-purged dns query ip responses
+type DnsCacheStorage struct {
+	cache *cache.Cache
+}
+
+func NewDnsCacheStorage(expiration, checkInterval time.Duration) *DnsCacheStorage {
+	storage := DnsCacheStorage{cache.New(expiration, checkInterval)}
+	return &storage
+}
+
+// Items returns map of non expired dns cache items
+func (dc *DnsCacheStorage) Items(includeExpired bool) map[string]DnsCacheItem {
+	var allItems = dc.cache.Items()
+
+	nonExpiredItems := map[string]DnsCacheItem{}
+
+	for k, v := range allItems {
+		if !includeExpired && v.Expired() {
+			continue
+		}
+		nonExpiredItems[k] = v.Object.(DnsCacheItem)
+	}
+
+	return nonExpiredItems
+}
+
+// Get returns non expired item from cache
+func (dc *DnsCacheStorage) Get(key string) (DnsCacheItem, bool) {
+	item, found := dc.cache.Get(key)
+	if !found {
+		return DnsCacheItem{}, false
+	}
+	return item.(DnsCacheItem), found
+}
+
+func (dc *DnsCacheStorage) Delete(key string) {
+	dc.cache.Delete(key)
+}
+
+// FetchItem returns list of ips from cache or resolves them and add to cache
+func (dc *DnsCacheStorage) FetchItem(hostName string) ([]string, error) {
+	if hostName == "" {
+		return nil, fmt.Errorf("hostName can't be empty. hostName=%v", hostName)
+	}
+
+	item, ok := dc.Get(hostName)
+	if ok {
+		logger.Debugf("Dns record was populated from cache: hostName=%q, addrs=%q", hostName, item.addrs)
+		return item.addrs, nil
+	}
+
+	addrs, err := dc.resolveDNSRecord(hostName)
+	if err != nil {
+		return nil, err
+	}
+
+	dc.Set(hostName, addrs)
+	return addrs, nil
+}
+
+func (dc *DnsCacheStorage) Set(key string, addrs []string) {
+	logger.Debugf("Adding dns record to cache: key=%q, addrs=%q", key, addrs)
+	dc.cache.Set(key, DnsCacheItem{addrs}, cache.DefaultExpiration)
+}
+
+// Clear deletes all records from cache
+func (dc *DnsCacheStorage) Clear() {
+	dc.cache.Flush()
+}
+
+func (dc *DnsCacheStorage) resolveDNSRecord(host string) ([]string, error) {
```

Comment:
what is purpose of this function? it looks like it is just alias for `net.LookupHost`

Reply:
>it looks like it is just alias for `net.LookupHost`
Yes, exactly. But in general it isn't and it's designed so. 
Why? 
Because of good-design principles(intention to separate business logic to separate logical methods) and because this method do some io(eg: connect to dns server and exchange messages with it), which should be mocked within unit tests. 
And due to the fact that currently unit tests aren't written well(by written well I mean that each "method-under-the-test" mocks it's dependencies using dependecyInjection pattern or mocking library and doesn't have a separate dns server for that mock) and someday they will, I left this method for this purpose(with respect to KISS principle).
Does that answer your question in case if your concern is design-redundancy?

Reply:
`net.LookupHost` direct call looks OK to me, it does what it says in name and nothing else. as to `resolveDNSRecord` - my 1st though was that it might do some logic inside connected with different types of DNS records (as it suggest in name) but it doesn't do anything specific to DNS-record. I would be ok to leave it here as is  but this is just extra function call with extra memory allocation. Thus code will likely be changed to direct call to `net.LookupHost` with next performance optimization.